### ISSUE TITLE
Second vat percentage is emptied if only first is edited in item bulk edit mode

### ISSUE THREAD

**Issue by jekkos** [jekkos/opensourcepos#381]
Issue exists since early version of ospos. Problem here is that all item taxes are deleted and thus if only one is edited second will be lost in bulk edit mode.

**Comment by jekkos** [comment]
I looked at this issue this morning. At first I thought it would be easy to fix but problem is that item taxes table uses a composite primary key. Once taxes are updated then all registered tax for that item are deleted. 

An option here would be to delete taxes based on the tax name + value that the user inputs in one of the two fields. If he however changes the tax name then 3 taxes can appear on an item, which will cause problems as well. 

So in this case a user should always update both taxes or none at all. MIght be best to enforce that in the form and warn the user about it.

**Comment by daN4cat** [comment]
What do you want to do with this issue? Move it to 3.0 or close?

Also we should test bulk edit in 2.4, I wouldn't that it's flaw as in 3.0.

**Comment by jekkos** [comment]
I'll do some testing on bulk update in 2.4 first and see which fixes need to go to 3.0. In the latter there shouldn't be too much change beside the form restyle.

**Comment by daN4cat** [comment]
I did a quick check with bulk edit and there is no way to avoid changing the supplier because missing a Do-Nothing like in 3.0. Outside that it seems not affecting other fields if nothing is changed.

@odiea can you please test bulk edit in 2.4 too and see if there is anything else outside what I already said?

**Comment by odiea** [comment]
On my side still issue if you do not enter a tax%. Supplier updates properly but Category does nothing.

**Comment by odiea** [comment]
Now I see what you mean with supplier. Yes it does remove supplier if none is selected on bulk update.

**Comment by daN4cat** [comment]
Probably in the input fields there should be a string in grey saying "No change" if nothing is typed in and of course do nothing.

A string like in sales/register suggesting to Scan or type the item name.

**Comment by ais-one** [comment]
Hi, I think I see what is causing the problem:

FILE: application\views\items\form_bulk.php

there was an error on Firebug about the with the category autocomplete:

Error: cannot call methods on autocomplete prior to initialization; attempted to call method 'http://127.0.0.1/opensourcepos/index.php/items/suggest_category'

This basically caused the form submit to have problem. Once I commented it out below, it worked ok.

```
$(document).ready(function()
{	
/*
    $("#category").autocomplete("<?php echo site_url('items/suggest_category');?>",`{max:100,minChars:0,delay:10});
   $("#category").result(function(event, data, formatted)
    {
    });
    $("#category").search();
*/
```

**Comment by jekkos** [comment]
@daN4cat the design of update tax fields is quite a bit flawed. This is because there is no unique id for each row. If you would leave the first tax field empty and only fill the second, then second will appear in the first field after reopen. 

We could disable the second field if the first isn't updated, and add a validation or warning to the second if the first is filled. We could add it to the final confirm message as well (which should be easier). 

A lot of work for a feature I don't really use.

**Comment by daN4cat** [comment]
To be honest I don't know in which case you want to use a second taxation.
I understand cases where you may have different level of taxation, e.g. childrenware is 5%, adultware is 20%.

Who uses the second taxation? Maybe in the bulk edit we can just leave the first? @odiea maybe you can tell us as US has more variability around tax.

**Comment by osworx** [comment]
> @daN4cat To be honest I don't know in which case you want to use a second taxation

For example here in Austria every Restaurant, Hotel and similiar.
Every business which has food and drinks or Bed & Breakfast has 2 taxes.

**Comment by daN4cat** [comment]
For the same item?

**Comment by osworx** [comment]
No, you asked for `taxation`.
Food = 20%
Drinks = 10%
But both on the same bill .. for example you go for Dinner.

**Comment by daN4cat** [comment]
Yes, but the issue here is to have one item having two taxations. I don't know where that is used.
The case to have one or another is known, as said earlier in my comment.

The only case is federal level Tax1, nation level Tax2 and the two are cumulative.

**Comment by osworx** [comment]
Okay, my fault. Thought discussion here is about more than one tax (on the same bill).
Having to taxes on the same item (product) is - I guess - nowhere happen in this world.
Either the one .. or the other, if there could be 2 (for the same item/product)

**Comment by jekkos** [comment]
This feature has been in there since the very beginning, so I guess it must be something american.

**Comment by daN4cat** [comment]
Only @odiea can tell us...

**Comment by odiea** [comment]
Yes here in the States we may have a State and a local tax. Most of the time they are just combined but if used it would be easier to do your monthly tax reports.

**Comment by jekkos** [comment]
I've done a fix here. Warning will be shown if first tax field is filled in. Also second will be disabled until first is filled.

**Comment by jekkos** [comment]
Hope this solution is acceptable for all.

**Comment by jekkos** [comment]
Would be good if someone can verify if ok so we can close.

**Comment by daN4cat** [comment]
There is a problem in 3.0 at least:

- items have a tax assigned
- nothing is changed
- tax is set to null with no warning

I my opinion it should behave like the prices, nothing is set and the field is left unchanged.
If somebody wants to change the tax it sets 0.

**Comment by daN4cat** [comment]
Also checking the code I can see that the bulk_edit form assumes a tax name & percent is passed, but the controller sets none.
The issue here is that in a bulk edit you don't know what is set or not.
Maybe it's better to set the config name and percent and let the user set to null one or the other to do nothing.

**Comment by daN4cat** [comment]
In 3.0 I fixed the issue:

- first set in the bulk edit the config defaults for the tax name and %
- if left with default, the default will be saved
- the string if emptied, the tax will not be changed

**Comment by daN4cat** [comment]
Fixed in 2.4:

- save category when set, do nothing when empty
- don't reset the supplier if left to none
- set the tax percentage only if not empty

**Comment by daN4cat** [comment]
I close because I believe this is now completed and I tested a bit and looks fine to me.
